src/electron/ipc/ipc-message-dispatcher.ts

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { TelemetryData } from 'common/extension-telemetry-events';
+import { Message } from 'common/message';
+import { ActionMessageDispatcher } from 'common/message-creators/types/dispatcher';
+import { IPC_MESSAGE_CHANNEL_NAME } from './ipc-channel-names';
+
+// The most common type of sink is BrowserWindow.webContext.send
+export type IpcMessageSink = (channelName: string, message: Message) => void;
+
+export class IpcMessageDispatcher implements ActionMessageDispatcher {
+    public dispatchMessage(message: Message): void {
+        if (this.registeredSinks.length === 0) {
+            this.messageBacklog.push(message);
+            return;
+        }
+
+        for (const sink of this.registeredSinks) {
+            sink(IPC_MESSAGE_CHANNEL_NAME, message);
+        }
+    }
+
+    public dispatchType(messageType: string): void {
+        throw new Error('Method not implemented (not yet required for any main process events).');
+    }
+    public sendTelemetry(eventName: string, eventData: TelemetryData): void {
+        throw new Error('Method not implemented (not yet required for any main process events).');
+    }
+
+    private messageBacklog: Message[] = [];
+    private registeredSinks: IpcMessageSink[] = [];
+
+    public registerMessageSink = (sink: IpcMessageSink): void => {
+        this.registeredSinks.push(sink);
+
+        if (this.registeredSinks.length === 1) {
```

Comment:
this `=== 1` here feels a little odd. I guess I kind of agree with your first gut reaction on having a queue-like behavior here: it would be better not to have it; that does raise the small concern of how do we ensure order of execution between instanciating this class and sending message. Then again, we own all the initializer so we can ensure that order.

Reply:
Going to remove the queue behavior and just control the instantiation order